frontend/javascripts/admin/dataset/dataset_add_compose_view.tsx

```diff
@@ -0,0 +1,423 @@
+          name: datasetName2,
+        }),
+      ]);
+      console.log("dataset1, dataset2", dataset1, dataset2);
+
+      const nodes1 = Array.from(
+        values(trees1)
+          .map((tree) => Array.from(tree.nodes.values())[0])
+          .values(),
+      );
+      const nodes2 = Array.from(
+        values(trees2)
+          .map((tree) => Array.from(tree.nodes.values())[0])
+          .values(),
+      );
+
+      for (const [node1, node2] of _.zip(nodes1, nodes2)) {
+        if ((node1 == null) != (node2 == null)) {
+          throw new Error("A tree was empty while its corresponding tree wasn't.");
+        }
+        if (node1 != null && node2 != null) {
+          sourcePoints.push(node1.position);
+          targetPoints.push(node2.position);
+        }
+      }
+      const datasets = [dataset1, dataset2];
+      setLinkedDatasets(datasets);
+      const newMatrix = estimateAffineMatrix4x4(sourcePoints, targetPoints);
+      const newLinks: LayerLink[] = (
+        _.flatMap(datasets, (dataset) =>
+          dataset.dataSource.dataLayers.map((layer) => [dataset, layer]),
+        ) as [APIDataset, APIDataLayer][]
+      ).map(
+        ([dataset, dataLayer]): LayerLink => ({
+          // todo: backend should expect datasetId
+          id: {
+            // todo: backend should expect owningOrganization, too
+            team: dataset.owningOrganization,
+            name: dataset.name,
+          },
+          datasetId: {
+            owningOrganization: dataset.owningOrganization,
+            name: dataset.name,
+          },
+          sourceName: dataLayer.name,
+          newName: dataLayer.name,
+          transformations:
+            dataset === datasets[0]
+              ? [
+                  {
+                    type: "affine",
+                    matrix: flatToNestedMatrix(newMatrix),
+                  },
+                ]
+              : [],
+        }),
+      );
+
+      form.setFieldsValue({ layers: newLinks });
+    }
+  };
+
+  const handleSubmit = async () => {
+    if (activeUser == null) {
+      throw new Error("Cannot upload dataset without being logged in.");
+    }
+    const layers = form.getFieldValue(["layers"]);
+
+    const uploadableDatastores = props.datastores.filter((datastore) => datastore.allowsUpload);
+    const datastoreToUse = uploadableDatastores[0];
+    if (!datastoreToUse) {
+      Toast.error("Could not find datastore that allows uploading.");
+      return;
+    }
+
+    const newDatasetName = form.getFieldValue(["name"]);
+    setIsLoading(true);
+    try {
+      await createDatasetComposition(datastoreToUse.url, {
+        newDatasetName,
+        targetFolderId: form.getFieldValue(["targetFolderId"]),
+        organizationName: "sample_organization",
+        scale: linkedDatasets[1].dataSource.scale,
+        layers,
+      });
+      await Utils.sleep(3000); // wait for 3 seconds so the server can catch up / do its thing
```

Comment:
@frcroth This code is similar to what is already happening after a dataset upload. After the dataset is imported, the user is asked whether they want to view the dataset. Without the delay, the dataset is usually "not there yet" and cannot be found. For the new compose case, I ran into the 404 error even with this 3s delay. I don't know what exactly causes this delay (maybe the communication with the datastore who reports things?).
Ideally, the upload/compose/add route would only finish once the dataset can also be found when viewing it. Maybe you could have a look at that? Don't know whether it's complicated.

cc @fm3

Reply:
That should be possible, let me know if you need help debugging what’s going on there.

Reply:
nice, seems to work now :tada:

Reply:
@frcroth Could you also add the `dataSourceService.checkInbox` code once the `/data/datasets/finishUpload` route finishes? Then, I can also remove the sleep in the upload case.